### ISSUE TITLE
misuse of env in eval(substitute) usage

### ISSUE THREAD

**Issue by dankelley** [dankelley/oce#1712]
RStudio's "code diagnostics" action shows as below. I am not clear on what `env` is; the `eval()` parameters are named `expr`, `envir` and `enclos`.  I will try naming them, to see if the diagnostic message goes away.

<img width="724" alt="Screen Shot 2020-07-12 at 1 05 59 PM" src="https://user-images.githubusercontent.com/99469/87251274-bd3c0600-c440-11ea-989d-ae94cc0e7654.png">

**Comment by dankelley** [comment]
The following patch shows how to silence the diagnostic message about `env`, for the first reported instance.  Before going on to change these calls globally, I will now check into the second message, about `fdef`.

```diff
diff --git a/R/AllClass.R b/R/AllClass.R
index 5a11f4fa..73135209 100644
--- a/R/AllClass.R
+++ b/R/AllClass.R
@@ -347,7 +347,7 @@ setMethod(f="subset",
           definition=function(x, subset, ...) {
               if (missing(subset))
                   stop("must give 'subset'")
-              keep <- eval(substitute(subset), x@data, parent.frame())
+              keep <- eval(expr=substitute(expr=subset, env=environment()), envir=x@data, enclos=parent.frame())
               ##message("percent keep ", round(sum(keep)/length(keep)*100, 2), "%")
               res <- x
               for (i in seq_along(x@data))
```

**Comment by dankelley** [comment]
The `fdef` is because
```R
standardGeneric("composite")
```
is not specifying the second arg, named `fdef`.  The weird thing is that the docs for `standardGeneric` say that this is never passed when defining a new generic.  The following patch makes this diagnostic message disappear, so I'll try a trial build with it, before doing this widely.  (Note that the function I'm giving is just an empty one.)

```diff
diff --git a/R/AllClass.R b/R/AllClass.R
index 5a11f4fa..7152e1d3 100644
--- a/R/AllClass.R
+++ b/R/AllClass.R
@@ -599,7 +599,7 @@ setMethod(f="show",
 #' @template compositeTemplate
 setGeneric("composite",
            function(object, ...) {
-               standardGeneric("composite")
+               standardGeneric("composite", function(x) x)
           })
```

**Comment by dankelley** [comment]
Hm, that thing I tried for `standardGeneric` does not work.  And the docs clearly imply that `fdef` is optional;  the following is a quote from `?standardGeneric`
> The argument fdef is inserted (automatically) when dispatching methods for a primitive function. If present, it must always be the function definition for the corresponding generic. Don't insert this argument by hand, as there is no validity checking and miss-specifying the function definition will cause certain failure.

and so I am going to ignore this diagnostic, and retain the previous form of my calls which, I might add, matches that seen in all my websearches, of which the most familiar is http://adv-r.had.co.nz/S4.html

**Comment by dankelley** [comment]
The `env` diagnostic is addressed in commit d610fb047e3d21fce9f638649e82994172278c2e of the "develop" branch.  I am going to ignore the `fdef` diagnostic, because it's a false positive, given that the documentation cited in the previous comment clearly states that it's OK not to supply `fdef`.  Also, a quick `grep` through the R source code did not turn up any cases where `fdef` was supplied (although I did not look terribly hard).

Perhaps `standardGeneric` is checking whether `fdef` has not been supplied.

**Comment by dankelley** [comment]
I'm closing this because the things that needed addressing have been adressed.
